Infrastructure/Services/ProcessEmailsService.cs

```diff
@@ -0,0 +1,450 @@
+                    Update the LastEmailSendAttemptDate = now, the status to BeenProcessed and LastStatusChangedDate = now
+                    Update the EmailNotifyLog with Notify response details and clear any previous ExceptionCode & ExceptionMessage - if retry then need to ensure that the Archive has previous notify info.
+                If error then log in EmailNotifyLog.ExceptionCode & EmailNotifyLog.ExceptionMessage
+                    Set status to ProcessingFailure and the ProcessFromDate = now.AddSeconds(RetrySendInSeconds)
+         */
+
+        var emailsSent = 0;
+
+        if (_emailSettingsConfig.SendEmailsFromNtp)
+        {
+            var emailsToSend = await _unitOfWork.EmailLogRepository
+                .GetAllAsync(x => (x.EmailStatus.AllowEmailSending &&
+                                    x.ProcessFromDate != null &&
+                                    x.ProcessFromDate <= DateTime.UtcNow &&
+                                    x.FinishProcessingDate > DateTime.UtcNow &&
+                                    (emailLogIds == null ||
+                                    emailLogIds.Count() == 0 ||
+                                    emailLogIds.Contains(x.Id))),
+                "EmailStatus,EmailPersonalisationLogs",
+                true);
+
+            if (emailsToSend.Any())
+            {
+                var notifyEmails = emailsToSend.Select(x => new NotifyEmailDto()
+                {
+                    Email = x.EmailAddress,
+                    EmailAddressUsedForTesting = x.EmailAddressUsedForTesting,
+                    ClientReference = x.ClientReferenceNumber,
+                    EmailTemplateType = x.EmailTemplateShortName.GetEnumFromDisplayName<EmailTemplateType>(),
+                    Personalisation = x.EmailPersonalisationLogs == null ?
+                        new Dictionary<string, dynamic>() :
+                        x.EmailPersonalisationLogs!.Select(y => new KeyValuePair<string, dynamic>(y.Key, y.Value)).ToDictionary(x => x.Key, x => x.Value)
+                }).ToList();
+
+                try
+                {
+                    await _notificationsClientService.SendEmailAsync(notifyEmails);
+                }
+                catch
+                {
+                    if (throwExceptions)
+                    {
+                        await UpdateSentEmailsAsync(emailsToSend, notifyEmails);
+                        throw;
+                    }
+                }
+
+                emailsSent = await UpdateSentEmailsAsync(emailsToSend, notifyEmails);
+            }
+        }
+        else
+        {
+            await SetEmailsToDeactivatedEmailsAsync();
+        }
+
+        return emailsSent;
+    }
+
+    private async Task<int> UpdateSentEmailsAsync(IEnumerable<EmailLog> emailsToUpdate, IEnumerable<NotifyEmailDto> emailResults)
+    {
```

Comment:
We should create separate methods for each function to improve the readability and maintainability of this method. We can move the `await _unitOfWork.Complete(); `out of the loop to save data to the Database after processing all emails, which will help us reduce the number of DB calls and improve performance.

Here is an example:

```
private async Task<int> UpdateSentEmailsAsync(IEnumerable<EmailLog> emailsToUpdate, IEnumerable<NotifyEmailDto> emailResults)
{
    NotifyResponseMappingConfig.Configure();

    int sentEmails = 0;
    DateTime updateSentDateTime = DateTime.UtcNow;

    List<EmailLog> emailsToUpdateList = new List<EmailLog>();

    foreach (var emailResult in emailResults)
    {
        EmailLog? emailToUpdate = null;
        try
        {
            emailToUpdate = emailsToUpdate.Single(x => x.ClientReferenceNumber == emailResult.ClientReference);

            UpdateEmailLogWithNotifyResponse(emailToUpdate, emailResult);

            int oldStatus = emailToUpdate.EmailStatusId;

            if (!string.IsNullOrEmpty(emailToUpdate.EmailNotifyResponseLog.NotifyId))
            {
                UpdateEmailLogStatusToBeenProcessed(emailToUpdate, ref sentEmails);
            }
            else
            {
                UpdateEmailLogStatusToProcessingFailure(emailToUpdate, updateSentDateTime);
            }

            UpdateEmailLogDates(emailToUpdate, updateSentDateTime, oldStatus);
            emailsToUpdateList.Add(emailToUpdate);
        }
        catch (Exception ex)
        {
            LogErrorInUpdateSentEmailsAsync(emailToUpdate, ex);
        }
    }

    // Update all email logs in the database at once
    foreach (var emailLog in emailsToUpdateList)
    {
        _unitOfWork.EmailLogRepository.Update(emailLog);
    }

    await _unitOfWork.Complete();

    return sentEmails;
}

private void UpdateEmailLogWithNotifyResponse(EmailLog emailToUpdate, NotifyEmailDto emailResult)
{
    var notifyResponse = emailToUpdate.EmailNotifyResponseLog ?? new EmailNotifyResponseLog();
    notifyResponse = emailResult.NotifyResponse.Adapt(notifyResponse);
    emailToUpdate.EmailNotifyResponseLog = notifyResponse;
}

private void UpdateEmailLogStatusToBeenProcessed(EmailLog emailToUpdate, ref int sentEmails)
{
    emailToUpdate.EmailStatusId = (int)EmailStatus.BeenProcessed;
    sentEmails++;
}

private void UpdateEmailLogStatusToProcessingFailure(EmailLog emailToUpdate, DateTime updateSentDateTime)
{
    emailToUpdate.EmailStatusId = (int)EmailStatus.ProcessingFailure;
    var retryInSeconds = _unitOfWork.EmailStatusRepository.GetById(emailToUpdate.EmailStatusId).RetrySendInSeconds;
    emailToUpdate.ProcessFromDate = updateSentDateTime.AddSeconds((double)retryInSeconds!);
}

private void UpdateEmailLogDates(EmailLog emailToUpdate, DateTime updateSentDateTime, int oldStatus)
{
    emailToUpdate.LastEmailSendAttemptDate = updateSentDateTime;
    if (oldStatus != emailToUpdate.EmailStatusId)
    {
        emailToUpdate.LastStatusChangedDate = updateSentDateTime;
    }
}

private void LogErrorInUpdateSentEmailsAsync(EmailLog? emailToUpdate, Exception ex)
{
    _logger.LogError(ex, "Unexpected error in UpdateSentEmailsAsync for EmailLog.ClientReferenceNumber: {ClientReferenceNumber}, EmailLog.Id: {EmailLogId}",
        emailToUpdate == null ? "NULL" : emailToUpdate.ClientReferenceNumber,
        emailToUpdate == null ? "NULL" : emailToUpdate.Id);
}

```

Reply:
I have split out `UpdateEmailLogWithNotifyResponse` as suggested, but the same as previous comment - not sure this is that long a method.

Same as previous comment too, need to save each record rather than as a unit - have added a comment to the code.